### ISSUE TITLE
Unsupervised Time Series Representation Learning

### ISSUE THREAD

**Issue by khirotaka** [khirotaka/enchanter#117]
Create a custom runner and a triplet loss for representation learning for time series.

## Motivation
Paper [Unsupervised Scalable Representation Learning for Multivariate Time Series](https://github.com/White-Link/UnsupervisedScalableRepresentationLearningTimeSeries) is a highly versatile technique in the time series field, but the portability of the original implementation code is low.

## Description
Enable unsupervised representation learning using Triplet Loss with only the following code.

```python
encoder = CNN(...)
runner = UnsupervisedTSRunner(...)
runner.add_loader("train", ...)
runner.train_config(...)
runner.run()
```

## Additional context (optional)
```python
loader.dataset.data
```

**Comment by khirotaka** [comment]
The loss function is defined as
![tripletloss](https://user-images.githubusercontent.com/24205015/91636274-3ff14280-ea3a-11ea-882c-3f6e1095b6a2.png)

It is easier to reuse if it is implemented by dividing it into "positive example vs anchor" and "negative example vs anchor" rather than `class TripletLoss`.

**Comment by khirotaka** [comment]
Negative examples of official implementation are as follows

https://github.com/White-Link/UnsupervisedScalableRepresentationLearningTimeSeries/blob/baad6d56c55f226ee3689243f94427d6687451f1/losses/triplet_loss.py#L133-L147

**Comment by khirotaka** [comment]
Defining some custom `Dataset` class

![algo](https://user-images.githubusercontent.com/24205015/92728160-3e653a00-f3ab-11ea-87e5-7265e24476c3.jpeg)

This algorithm has the following problems.
* _x^{ref}_ and _x^{neg}_ may be selected from the same sample.
* _i_k_ should be a separate sample from _i_.

As a solution
1. Use class labels to ensure that x ^{reg} and x ^{neg} are in different classes.
2. Assign an ID to each sample in the dataset so that _i_k_ and _i_ are separate samples.

```python
class LabeldDataset(Dataset):
    def __init__(self, data, target):
        self.data = data
        self.target = target
    ....

```

```python
class UnsupervisedDataset(Dataset):
    def __init__(self, data):
        self.data = data
        self.ids = [i for i in range(len(data))
    ....
```

**Comment by khirotaka** [comment]
Early Stopping
```python
class EarlyStoppingForUSTS(Callback):
    def __init__(self, classifier, data, target, monitor="accuracy", kfold=None):
        super().__init__()
        self.classifier = classifier
        self.data = data
        self.target = target
        if kfold:
            self.kfold = fold
        else:
            self.kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
        self.monitor = monitor
        self.device = None

    def encode(self):
        out = []
        self.model.eval()
        with torch.no_grad():
            self.data = self.data.to(self.device)
            out.append(self.model(data).cpu().numpy())
        return out

    def cross_val(self):
        features = self.encode()
        score = cross_validate(self.classifier, features, self.target, cv=self.kfold, scoring=self.monitor)["test_{}".format(self.monitor)].mean()
        return score

    def on_epoch_end(self, epoch, logs: Optional[Dict] = None) -> None:
        ...
```

**Comment by khirotaka** [comment]
1. create `evaluator` on `__init__`
https://github.com/khirotaka/enchanter/blob/e5ea15424384ccaa8b049efdbb961c21e357f050/enchanter/tasks/ts_unsupervised.py#L30-L41

2. update `.encode()`
https://github.com/khirotaka/enchanter/blob/e5ea15424384ccaa8b049efdbb961c21e357f050/enchanter/tasks/ts_unsupervised.py#L195-L208

```python
def encode(self, data: Union[torch.Tensor, np.ndarray]) -> Union[torch.Tensor, np.ndarray]:
    # When np.ndarray is input, output np.ndarray. When torch.Tensor is input, output torch.Tensor.
```


3. override `.val_step()`, `.val_end()`, `.test_step()`, `.test_end()`
```python
def val_step(self, batch):
    x, y = batch
    output = self.encode(x)
    return {“encoded”: output, “targets”: y}

def val_end(self, output):
    outputs = torch.cat( ... ).cpu().numpy()
    targets = torch.cat( ... ).cpu().numpy()

    logs = cross_validate(self.evaluator, outputs, targets)
    return {“logs”: logs}
...

**Comment by khirotaka** [comment]
https://github.com/khirotaka/enchanter/blob/323424fe01d896f07172f34044a30a1d07d7d129/enchanter/tasks/ts_unsupervised.py#L95

change method name `calculate_negative_loss_per_negative_sample ` -> `calculate_negative_loss_per_sample`
